verifier/src/main/kotlin/tech/figure/classification/asset/verifier/util/eventstream/providers/BlockApiEventStreamProvider.kt

```diff
@@ -0,0 +1,89 @@
+package tech.figure.classification.asset.verifier.util.eventstream.providers
+
+import io.provenance.client.protobuf.extensions.time.toOffsetDateTimeOrNull
+import io.provenance.eventstream.stream.models.Event
+import io.provenance.eventstream.stream.models.TxEvent
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.isActive
+import tech.figure.block.api.client.BlockAPIClient
+import tech.figure.block.api.proto.BlockServiceOuterClass
+import tech.figure.classification.asset.verifier.config.EventStreamProvider
+import tech.figure.classification.asset.verifier.provenance.AssetClassificationEvent
+
+class BlockApiEventStreamProvider(
+    private val blockApiClient: BlockAPIClient,
+    private val coroutineScope: CoroutineScope,
+    private val batchSize: Long = 100L,
+) : EventStreamProvider {
+
+    override suspend fun currentHeight(): Long =
+        blockApiClient.status().currentHeight
+
+    override suspend fun startProcessingFromHeight(
+        height: Long?,
+        onBlock: suspend (blockHeight: Long) -> Unit,
+        onEvent: suspend (event: AssetClassificationEvent) -> Unit,
+        onError: suspend (throwable: Throwable) -> Unit,
+        onCompletion: suspend (throwable: Throwable?) -> Unit
+    ) {
+
+        val currentHeight = currentHeight()
+        var startingHeight = height?.let {
+            if (it > currentHeight) currentHeight else it
+        } ?: getStartingBlockHeight(currentHeight)
+
+        while (coroutineScope.isActive) {
+
+            (startingHeight..(startingHeight + batchSize))
+                .forEach { blockHeight ->
+                    runCatching {
+                        blockApiClient.getBlockByHeight(
```

Comment:
Kind of a bummer you have to hit the API on a per-block basis.  Is there some form of streaming or block range query you can do?  I haven't played with block api like at all so I'm not sure if that's a thing, but this seems like the client would really hammer the service

Reply:
Not that im aware of :/ I can reach out to Phil and see if theres a better way to do this, but as of today theres nothing in the blockapi protos to support querying a range of block AFAIK.